### PR TITLE
Fix existing tests and add new tests

### PR DESCRIPTION
**Pull request by desilinguist**
This PR improves the testing and increases the test coverage from 62% to 78%.

1. It fixes `test_parse_util.py` which was not running with `nosetests` before. However, since the ZPar models do not fit in the Travis worker's memory, this test is run on our internal Bamboo CI and skipped on Travis.

2. It adds `test_tree_util.py` for two of the functions in `tree_util.py` that Mike had suggested to add unit tests for.

3. It adds `test_rst_parse.py` which tests the `segment_and_parse()` function in `rst_parse.py`. One of those tests is skipped on Travis because of the ZPar models being too large and run only on our internal Bamboo CI. Adding these tests exposed a bug in this function which is now fixed. 

4. It adds another tests to `test_discourseparsing.py` that Mike had suggested.

5. Adds a brand new end-to-end test that basically trains a segmenter and RST parser on the RST Discourse Treebank.  Basically, this runs all the steps in the README. This only runs on our internal Bamboo CI since it requires access to the treebank. 

6. It moves `try_head_rules.py` from `tests` to under `utils` and adds a new entry point for it to `setup.py`.

7. It uses the full path to `crf_test` and `crf_learn` where they are used since otherwise it doesn't work properly in the tests. 

8. Some other minor changes (replacing `os.path` with `pathlib`, docstring & comment changes).

9. Adds some new test data and models.

### CODE REVIEW

File: tests/test_rst_parse.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+        # make sure there are the expected number of EDUs
+        # and check the first and last EDUs as well
+        eq_(len(edu_tokens), 5)
+        eq_(edu_tokens[0], ['Rhetorical', 'structure', 'theory'])
+        eq_(edu_tokens[-1], ['and', 'defined', 'in', 'a', 'seminal',
+                             'paper', 'in', '1988', '.'])
+
+    def test_segment_and_parse_no_text(self):
+        """Test that parse pipeline fails gracefully with no text."""
```
Comment (ghost) [inline tests/test_rst_parse.py]: does it fail or just does not do anything?

File: tests/test_rst_parse.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        with open(self.partial_doc_dict_file, "r") as docdictfh:
+            doc_dict = json.load(docdictfh)
+
+        edu_tokens, rst_trees = segment_and_parse(doc_dict,
+                                                  self.parser,
+                                                  self.segmenter,
+                                                  self.rst_parser)
+
+        # check that the document dictionary now has other fields populated
+        ok_("edu_start_indices" in doc_dict)
```
Comment (ghost) [inline tests/test_rst_parse.py:102]: can we check these fields are not present in `doc_dict` before we call `segment_and_parse`? that catches unknowingly updating the test data later on.

**Review by ghost — DISMISSED** [review]
This looks great. I have two minor comments.

File: tests/test_rst_parse.py (hunk 1 above)
Comment (desilinguist) [reply tests/test_rst_parse.py]: Fair enough :)

File: tests/test_rst_parse.py (hunk 2 above)
Comment (desilinguist) [reply tests/test_rst_parse.py:102]: Sorry, I am not sure what you mean. Can you give me an example?

Comment (desilinguist) [reply tests/test_rst_parse.py:102]: Here's what @bndgyawali meant: what if someone adds that field to partial_doc_dict.json ? So we should also check that it wasn't there before we call `segment_and_parse()`?

Comment (ghost) [reply tests/test_rst_parse.py:102]: right.

**Review by aoifecahill — APPROVED** [review]
Looks good, thanks!

**Comment by desilinguist** [comment]
@aoifecahill you forgot to approve so I can't merge :)

**Review by ghost — APPROVED** [review]

**Pull request merged**